### PR TITLE
[DependencyInjection] Ensure AutowireCallable handles non-shared services

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | #51691
| License       | MIT
| Doc PR        | -

@miqrogroove can you please confirm this fixes the issue for you?

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -1062,9 +1062,17 @@ private function createService(Definition $definition, array &$inlineServices, b
                 $callable[0] instanceof Reference
                 || $callable[0] instanceof Definition && !isset($inlineServices[spl_object_hash($callable[0])])
             )) {
-                $initializer = function () use ($callable, &$inlineServices) {
-                    return $this->doResolveServices($callable[0], $inlineServices);
-                };
+                if ($callable[0] instanceof Definition && $callable[0]->isShared()) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php:1065]: if `$callable[0]` is a reference, shouldn't we also check whether this reference targets a shared service ?

File: src/Symfony/Component/DependencyInjection/Argument/LazyClosure.php [hunk 2]
```diff
@@ -25,26 +24,9 @@
  */
 class LazyClosure
 {
-    public readonly object $service;
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Argument/LazyClosure.php:28]: isn't it a BC break to change the property from public to protected ?

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php:1065]: if it's a reference, we already dump code that does the memoization

File: src/Symfony/Component/DependencyInjection/Argument/LazyClosure.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Argument/LazyClosure.php:28]: `@internal` to the rescue ;)

**Comment by miqrogroove** [comment]
One of us broke something.

```
Cannot autowire service "App\Controller\MyController": argument "$myCallable" of method "__construct()" references class "Closure" but no such service exists.
```

I haven't changed the property syntax so...

```
    public function __construct(
        #[AutowireCallable(service: 'App\NonShared\MyService', method: 'myMethod')]
        private \Closure $myCallable,
    ) {
    }
```

I skipped the Test patches.  I was hoping it would just work 🙂

**Comment by nicolas-grekas** [comment]
Do you have the corresponding `use` statement for the attribute?

**Comment by miqrogroove** [comment]
Adding a `use` statement doesn't change anything for the Fully Qualified Name of `\Closure`.  I tried it just to make sure.  No difference.

**Comment by miqrogroove** [comment]
Sorry I rushed that test 😊 The attribute's `use` was needed.

Now, it seems like I'm still getting the same instance for each call to the closure...

**Comment by nicolas-grekas** [comment]
Can you please share a reproducer, or better: a test case?

**Comment by miqrogroove** [comment]
Yes.  If it takes a while to make a clean test case then I will follow up tonight.

**Comment by stof** [comment]
Personally, I think the behavior described in https://github.com/symfony/symfony/issues/51691#issuecomment-1725873377 by @derrabus would make sense.

So in your test, if you have a single service in which you use AutowireCallable on a single argument, it would still be expected that the same instance is used.

**Comment by miqrogroove** [comment]
Testcase https://github.com/miqrogroove/symfony-bug-51691/tree/main/trunk

`bin/console app:test`

**Comment by nicolas-grekas** [comment]
Thanks. Did you forget to commit some files? There's no bin/console file there :)

**Comment by miqrogroove** [comment]
> So in your test, if you have a single service in which you use AutowireCallable on a single argument, it would still be expected that the same instance is used.

It would be a documentation issue then.  If that's the conclusion then the PR wouldn't be needed.

**Comment by miqrogroove** [comment]
> Thanks. Did you forget to commit some files? There's no bin/console file there :)

Added bin/console.

**Comment by nicolas-grekas** [comment]
Thanks. So we currently generate something like this:
```php
$container->privates['App\\Command\\TestCommand'] = $instance = new \App\Command\TestCommand((new \App\NonShared\MyService())->setter(...))
```

This is the expected behavior to me. The service is non-shared with other services, but it's still the same one that is reused by individual consumer services when used as a callable.

A service-closure is different in the sense that it's a sort of factory for a service, so it's expected that it returns a new instance for non-shared ones.

Closing therefor, thanks for the help understanding this.

**Pull request closed without merging**